xonsh/shell.py

```diff
@@ -121,54 +121,36 @@ class Shell(object):
         "best": "best",
         "d": "dumb",
         "dumb": "dumb",
-        "ptk": "prompt_toolkit",
-        "ptk1": "prompt_toolkit1",
-        "ptk2": "prompt_toolkit2",
+        "ptk": "prompt_toolkit",  # there's only 1 prompt_toolkit shell (now)
+        "ptk1": "prompt_toolkit",  # allow any old config reference to use it
+        "ptk2": "prompt_toolkit",  # so long as user actually  has ptk2+ installed.
         "prompt-toolkit": "prompt_toolkit",
         "prompt_toolkit": "prompt_toolkit",
-        "prompt-toolkit1": "prompt_toolkit1",
-        "prompt-toolkit2": "prompt_toolkit2",
+        "prompt-toolkit1": "prompt_toolkit",
+        "prompt-toolkit2": "prompt_toolkit",
         "rand": "random",
         "random": "random",
         "rl": "readline",
         "readline": "readline",
     }
 
-    def __init__(self, execer, ctx=None, shell_type=None, **kwargs):
-        """
-        Parameters
-        ----------
-        execer : Execer
-            An execer instance capable of running xonsh code.
-        ctx : Mapping, optional
-            The execution context for the shell (e.g. the globals namespace).
-            If none, this is computed by loading the rc files. If not None,
-            this no additional context is computed and this is used
-            directly.
-        shell_type : str, optional
-            The shell type to start, such as 'readline', 'prompt_toolkit1',
-            or 'random'.
-        """
-        self.execer = execer
-        self.ctx = {} if ctx is None else ctx
-        env = builtins.__xonsh__.env
-        # build history backend before creating shell
-        builtins.__xonsh__.history = hist = xhm.construct_history(
-            env=env.detype(), ts=[time.time(), None], locked=True
-        )
-
+    @staticmethod
+    def choose_shell_type(shell_type=None, env={}):
         # pick a valid shell -- if no shell is specified by the user,
         # shell type is pulled from env
+        # extracted for testability
         if shell_type is None:
             shell_type = env.get("SHELL_TYPE")
             if shell_type == "none":
                 # This bricks interactive xonsh
                 # Can happen from the use of .xinitrc, .xsession, etc
+                # odd logic.  We don't override if shell.__init__( shell_type="none"),
+                # only if it come from environment?
                 shell_type = "best"
-        shell_type = self.shell_type_aliases.get(shell_type, shell_type)
+        shell_type = Shell.shell_type_aliases.get(shell_type, shell_type)
         if shell_type == "best" or shell_type is None:
             shell_type = best_shell_type()
-        elif env.get("TERM", "") == "dumb":
+        elif env.get("TERM") == "dumb":
```

Comment:
I think the string-string comparison is faster than the None-string compare:

```
scopatz@artemis ~/downloads/t $ timeit! "" == "dumb"                                                                                                                   
10000000 loops, best of 3: 20.5 ns per loop
scopatz@artemis ~/downloads/t $ timeit! None == "dumb"                                                                                                                 
10000000 loops, best of 3: 24.7 ns per loop
```

And since this is a startup task, it is important to try to be as fast as possible :wink:

Reply:
OK, diddit. I'd argue more about it, but our time is precious.